### ISSUE TITLE
Pre-/post-backup scripts are only run for "borg create", not for prune, list or check

### ISSUE THREAD

**Issue by frederikmoellers** [borgbase/vorta#379]
Somewhere between 0.6.10 and 0.6.22 vorta introduced a change that made sure that pre- and post-backup commands are run before Repo checks. I think this was PR #264. This enabled me to use pre- and post-backup scripts for mounting and unmounting a samba share where the actual backup is stored.

However, these commands seem to only run when an actual backup task is being executed (`borg create`). Most importantly, it seems that the post-backup command is run after the backup finished but *before* the pruning is done. This results in the prune command failing (repo not found) and thus no pruning happening at all.

In vorta's GUI log I see that the returncodes are nonzero whenever the command tries to access the repo, but not when performing a backup via `borg create`:
![image](https://user-images.githubusercontent.com/6661317/67945102-6823d100-fbde-11e9-830f-183845d5e429.png)

In the log file I see that the post-backup command seems to run directly after `create`, but before `prune`. Furthermore, I see that all commands except `borg create` seem to fail:
```
2019-10-31 12:00:00,793 - apscheduler.executors.default - INFO - Running job "VortaScheduler.create_backup (trigger: cron[hour='12', minute='0'], next run at: 2019-10-31 12:00:00 CET)" (scheduled at 2019-10-31 12:00:00+01:00)
2019-10-31 12:00:00,797 - vorta.scheduler - INFO - Starting background backup for Default
2019-10-31 12:00:00,799 - vorta.notifications - DEBUG - success notifications suppressed
2019-10-31 12:00:00,809 - vorta.borg.borg_thread - DEBUG - Using VortaDBKeyring keyring to store passwords.
2019-10-31 12:00:02,688 - vorta.scheduler - INFO - Preparation for backup successful.
2019-10-31 12:00:02,718 - vorta.borg.borg_thread - INFO - Running command /usr/bin/borg create --list --info --log-json --json --filter=AM -C lz4 --exclude-if-present .nobackup /backup/Default/::computername-default-2019-10-31T12:00:02 /home/username/
2019-10-31 12:00:04,233 - vorta.borg.borg_thread - INFO - Creating archive at "/backup/Default/::computername-default-2019-10-31T12:00:02"
2019-10-31 12:01:22,933 - vorta.notifications - DEBUG - success notifications suppressed
2019-10-31 12:01:22,933 - vorta.scheduler - INFO - Backup creation successful.
2019-10-31 12:01:22,934 - vorta.scheduler - INFO - Doing post-backup jobs for Default
2019-10-31 12:01:22,936 - vorta.borg.borg_thread - DEBUG - Using VortaDBKeyring keyring to store passwords.
2019-10-31 12:01:22,961 - vorta.borg.borg_thread - INFO - Running command /usr/bin/borg prune --list --info --log-json --keep-hourly 0 --keep-daily 7 --keep-weekly 4 --keep-monthly 6 --keep-yearly 2 --prefix computername-default- --keep-within 10H /backup/Default/
2019-10-31 12:01:23,483 - vorta.borg.borg_thread - ERROR - Repository /backup/Default does not exist.
2019-10-31 12:01:23,646 - vorta.borg.borg_thread - DEBUG - Using VortaDBKeyring keyring to store passwords.
2019-10-31 12:01:23,676 - vorta.borg.borg_thread - INFO - Running command /usr/bin/borg list --info --log-json --json /backup/Default/
2019-10-31 12:01:24,064 - vorta.borg.borg_thread - ERROR - Repository /backup/Default does not exist.
2019-10-31 12:01:24,228 - vorta.borg.borg_thread - DEBUG - Using VortaDBKeyring keyring to store passwords.
2019-10-31 12:01:24,258 - vorta.borg.borg_thread - INFO - Running command /usr/bin/borg check --info --log-json /backup/Default/
2019-10-31 12:01:24,639 - vorta.borg.borg_thread - ERROR - Repository /backup/Default does not exist.
2019-10-31 12:01:24,791 - vorta.scheduler - INFO - Finished background task for profile Default
2019-10-31 12:01:24,793 - apscheduler.executors.default - INFO - Job "VortaScheduler.create_backup (trigger: cron[hour='12', minute='0'], next run at: 2019-11-01 12:00:00 CET)" executed successfully
```

**Comment by m3nu** [comment]
According to the logs, there is an issue with your repo path. Maybe a drive not mounted? Does it work with the CLI directly? (see the logs for the precise arguments)

`ERROR - Repository /backup/Default does not exist.`

**Comment by frederikmoellers** [comment]
Yes, the issue path is an empty directory. I use the pre- and post-backup scripts to mount a samba share into the directory. The backup is then on the share. The pre-backup command checks if the server is available and mounts the share. The post-backup command unmounts the share.

Unfortunately, the execution of the pre- and post-backup scripts does not seem to appear in the log, so this is not visible. However, judging from the log messages, I suppose the execution order is like this:

1. Run pre-backup script (mount samba share)
1. `borg create` (succeeds b/c share is mounted and repo is available)
1. Run post-backup script (unmount samba share)
1. `borg prune` (fails b/c share is no longer mounted and repo cannot be found)
1. `borg list` (fails for same reason)

Imho the post-backup script should only be run after *all* backup-related commands have been executed, so after the execution of `borg list`.

**Comment by frederikmoellers** [comment]
I just confirmed this in the code:

In [src/vorta/borg/create.py#L99](https://github.com/borgbase/vorta/blob/master/src/vorta/borg/create.py#L99), the pre-backup command is executed. When the thread ends (read: after `borg create` finishes), [#L46](https://github.com/borgbase/vorta/blob/master/src/vorta/borg/create.py#L46) executes the post-backup command.

However, this only happens for `create`, not for any of `prune` or `list`. In [prune.py](https://github.com/borgbase/vorta/blob/master/src/vorta/borg/prune.py), the pre- and post-backup commands are not executed. I think it should be possible to also have them execute for these subcommands.

If anyone can confirm that this would be useful, I can make a PR.

**Comment by Janhouse** [comment]
Having the same issue here.

**Comment by m3nu** [comment]
> If anyone can confirm that this would be useful, I can make a PR. Maybe it makes sense to execute the pre-backup script in [BorgThread.prepare()](https://github.com/borgbase/vorta/blob/master/src/vorta/borg/borg_thread.py#L87)? That would make sure that it's executed before any repository-related task.

Originally those pre/post-backup commands were meant to prepare the data for backup. Like e.g. Borgmatic does with their hooks.

If you need a command to even connect to the repo, I would regard it as different feature?

What are you guys using it for roughly? Probably not to prepare some data for backup (e.g. dump a database).

**Comment by Janhouse** [comment]
I use pre-backup/post-backup to mount and umount network storage. It is available only on certain networks so it can't be mounted all the time.
And if it is not mounted before backup or prune jobs, it won't work.

It should run before any task that involves using the repository, and run the post-backup task afterwards.

**Comment by m3nu** [comment]
This is quite different from the current use case, which is like a `before_create`/`after_create` hook. You are suggesting `before/after_everything`. How would this look in the UI?

**Comment by Janhouse** [comment]
It could be either a different set of pre/post scripts or maybe a checkbox next to existing ones with text "run before/after everything" and an argument passed to the script to know what event just happened.

**Comment by frederikmoellers** [comment]
I guess an argument to the scripts would be easiest and least intrusive. It doesn't need any GUI changes and lets the script decide what to do depending on the context: DB dump scripts will only act on "backup" arguments, mount scripts for network shares might act on any action.

**Comment by m3nu** [comment]
> a checkbox next to existing ones with text "run before/after everything" and an argument passed to the script to know what event just happened.

I can also live with a checkbox.

**Comment by jchhenderson** [comment]
yep, I have just come across this problem - I thought it was me for ages (pruning not running because post-backup script to dismount drive had been executed beforehand).

Is there timing for this?

**Comment by frederikmoellers** [comment]
> Is there timing for this?

Unfortunately, no. I will try to dedicate some time to this, but I can't give any estimates at the moment. If anyone is interested in working on this, feel free.

**Comment by frederikmoellers** [comment]
I have a PR almost ready with the functionality. One question rose, though:

Would it be useful to execute the commands on *every* action? This would even include running the post-backup-command after "mount", the pre-backup-command before "umount" and both commands around "version". For cases where the commands are used to e.g. mount filesystems, this doesn't make sense. `borg --version` can run without access to the repos and `borg mount` should not finish with unmounting the underlying filesystem (in the post-backup-command) where the repo is located (this would break the borg-mount).

The only use case I can imagine is if the commands are used to make the borg executable available in the first place. But I don't know if this is really realistic. Then again, it would be most consistent to run the commands on every action and let the scripts decide when to execute what (based on the `$subcommand` environment variable that tells them what is being done).

So I can't decide between being consistent and more friendly for weird side-cases (execute always) or being more friendly to most actual use cases (execute only where it makes sense, e.g. not before/after version).

**Comment by sten0** [comment]
Would it be reasonable to consider this an "advanced user feature", with richer support for hooks, and not clutter the gui with toggle buttons that may confuse users who don't need the feature?  What I mean is:

1. Vorta calls the pre and post hook scripts for every operation
2. However, it calls them with the one of the following arguments: check, create, list, prune
3. For user-friendliness, maybe provide an example script in the repository that just emits debugging echoes
4. Given this design, it's also possible to further simplify the UI, because
5. Vorta could call the hook script with two arguments
    * eg: `hook_script.sh pre create` and `hook_script.sh post create`
    * meaning there only needs to be one "hook script" path box.

The reason I think multiple cases should be supported and passed to the script as an argument is because `create` hooks are useful for things like snapshotting, but snapshot creation and destruction is not necessary or desired for check, list, nor prune.  I'm also not sure if Apple has completely walled-off APFS snapshot support and if `tmutil` (Time Machine infrastructure) is the only way to interact with them.  At any rate, that's a tangential issue to this one; although, it introduces the question: what if the source paths don't have stable names?  Time Machine snapshots have names like `com.apple.TimeMachine.2019-02-23-102421`, where 102421 appears to be some kind of transaction ID.  Of course, if Apple users don't have the ability to snapshot a consistent database state (eg: all those Photos and Music databases) then we can "Har!  Har!  Linux is better", but that doesn't seem very nice to me ;-)  Anyways, hypothetically, it seems like `hook_script.sh pre create` could return a path to `com.apple.TimeMachine.2019-02-23-102421`...but I suspect @m3nu might say that supporting this case is outside the scope of Vorta.

@frederikmoellers
> Would it be useful to execute the commands on _every_ action? This would even include running the post-backup-command after "mount", the pre-backup-command before "umount" and both commands around "version". For cases where the commands are used to e.g. mount filesystems, this doesn't make sense. `borg --version` can run without access to the repos and `borg mount` should not finish with unmounting the underlying filesystem (in the post-backup-command) where the repo is located (this would break the borg-mount).
> 
> The only use case I can imagine is if the commands are used to make the borg executable available in the first place. But I don't know if this is really realistic. Then again, it would be most consistent to run the commands on every action and let the scripts decide when to execute what (based on the `$subcommand` environment variable that tells them what is being done).
> 
I'm also not sure how often this ability would be used, but it seems like users who installed Vorta from PyPI might use it for things like activating the virtualenvironment.
> So I can't decide between being consistent and more friendly for weird side-cases (execute always) or being more friendly to most actual use cases (execute only where it makes sense, e.g. not before/after version). I could use some opinions on this.

The included example script could handle the `borg --version` pre and post cases using the `*` case, which should be a noop.  Users who want to do something exotic can write a function or copy their code into the `*` case, or define a `version` case.  To be extra user-friendly I guess the example script could provide a case for `borg --version`, or may just a comment.  There's also the question of support-burden for more cases.  For future compatibility reference during upgrades it might also be nice to see a quick-reference list of supported Vorta operations/cases near the top of the file.  In a way it's the eternal question of "how much can we expect others to infer meaning" vs "explicit comments and documentation".

Other than that, I wonder about error handling, and what Vorta (and the example script) should do if a single command in the called case fails.  Oh, and finally, logging considerations!  It seems to me that each command in the example script should be logged by Vorta.  @m3nu, what do have to say about design considerations?  The only other things I can think of is that the example script should be in POSIX SH and not BASH, and that it should be written knowing that many users will be running it suid root or with setcap (to get a subset of CAP_ADMIN).  Oh and there's also the question of what should happen if a user's hook script doesn't have a `*` case and doesn't handle a case (for whatever reasons).  Should it error, warn, or silently noop?

And @frederikmoellers, thank you for working on this!  P.S. I was thinking about working on this post-Debian 11 release (probably September'ish), but I'm happy to hear someone else has prioritised it :-)

**Comment by frederikmoellers** [comment]
Wow, that's a lot of (unexpected but very welcome) feedback :) Thanks a lot for giving this so much thought!

I really like the idea of having a simple UI with only one input box and leaving the configuration in the script to be done by advanced users. This gives us all the flexibility we might ever need without (as you say) cluttering the UI for regular users.

> Anyways, hypothetically, it seems like hook_script.sh pre create could return a path to com.apple.TimeMachine.2019-02-23-102421

That's an interesting use case. This approach certainly opens the possibility of e.g. pre-create hooks returning information which vorta/borg then uses for the backup. However, implementing this feedback channel seems like an awful lot of work and I'm not sure if there's a larger target audience for this feature. I recommend gathering some feedback before proceeding with this.

> I'm also not sure how often this ability would be used, but it seems like users who installed Vorta from PyPI might use it for things like activating the virtualenvironment.

Not sure I understand what you mean. Wouldn't they have to activate the virtualenv before starting vorta in the first place? Or do you mean users who installed *borg* from PyPI? Anyway, I do agree power users will eventually come up with a use case for this and given your approach (script is always called with 2 parameters; default behaviour is to pass and quit), it really makes no sense to not call it on every operation.

On the question of supported cases/subcommands: I suggest we just pass the subcommand to the script the same way it appears in vorta's (GUI) log (`check`/`list`/`prune`/`create`/`--version`). That way things should stay intact even if more subcommands become available. In the script I'd give examples for subcommands at the top and try to provide an extensive list, but always refer to borg and/or the vorta log for absolute certainty.

On error handling: I suggest we handle errors the same way we do now. If any script call returns with a code ≠0, we abort there and cancel the operation. Anything else is too complicated and error-prone imho. Whether the script itself aborts on every error is then left to the user. They can do elaborate error-handling or just abort on the first error as well.

I don't have a strong opinion on logging. Logging everything (stdout and stderr of the script and, consequently, of all programs called within) could clutter the log, but at the same time it would be the most user friendly option imho. So unless anyone disagree's, I'm going to prepare the PR to log everything. After all, users can always redirect to /dev/null in their script :)

I'll try to write a good example using POSIX SH and to make sure that users understand how it's going to be executed and what happens if they remove cases.

> And @frederikmoellers, thank you for working on this! P.S. I was thinking about working on this post-Debian 11 release (probably September'ish), but I'm happy to hear someone else has prioritised it :-)

Well what can I say… being directly affected and needing a feature is always the best motivation ;) At the moment the need is gone for me (which is why I haven't put any pressure on finishing the PR) but with your feedback I'll try to get this done now.

**Comment by m3nu** [comment]
Yeah, good idea to move it into scripts. Should those scipts live in the settings folder or a user-defined path?

One may want different scripts based on the profile? Or a way to temporarily disable them from the UI.

In addition to passing the subcommand as argument, it could pass more details as env vars. E.g. `VORTA_PROFILE_NAME`, `VORTA_REPO_URL`, ... Then the docs would have a sample script on how to deal with different subcommands?

```
case $1 in
   pre_create)
      # do stuff before borg create
      ;;
   post_create)
      # after borg create
      ;;
   pre_prune)
      # before borg prune
      ;;
   *)
     Default condition to be executed
     ;;
esac
```

**Comment by Erwyn** [comment]
Hi there,

is the thread up to date regarding this feature? I tried to check the documentation but couldn't find any mention of it.

**Comment by frederikmoellers** [comment]
Well the thread is up to date, but unfortunately there's no PR yet. I have to admit I haven't looked into this for quite a while since I stopped doing backups on network mounts, but I will get back to this now. You can expect a PR in a few days and there we can have a final discussion on where to put the scripts and the other open questions.

**Comment by real-yfprojects** [comment]
The hook names could look like the ones from borgmatic.

> Yeah, good idea to move it into scripts. Should those scripts live in the settings folder or a user-defined path?

The user could specify the file path.

> One may want different scripts based on the profile? Or a way to temporarily disable them from the UI.

A per profile entry would allow for multiple scripts while not forcing multiple. Some settings regarding scripts could be useful. 

> In addition to passing the subcommand as argument, it could pass more details as env vars. E.g. `VORTA_PROFILE_NAME`, `VORTA_REPO_URL`, ... Then the docs would have a sample script on how to deal with different subcommands?

Isn't that already done for the current scripts?
The hook name could also be passed as a env var so that the user can decide whether to pass it as an argument to the script.

**Comment by sten0** [comment]
Frederik Möllers ***@***.***> writes:

>  Wow, that's a lot of (unexpected but very welcome) feedback :) Thanks a lot for giving this so much thought!
>

Thanks! :-)  I'm happy you like this approach.  Sorry I missed your
reply until now!

>> Anyways, hypothetically, it seems like hook_script.sh pre create could return a path to com.apple.TimeMachine.2019-02-23-102421
>
> That's an interesting use case. This approach certainly opens the possibility of e.g. pre-create hooks returning information which vorta/borg then uses for the backup. However, implementing this feedback channel seems like an awful lot of work and I'm not sure if there's a larger target audience for this feature. I recommend gathering some feedback before proceeding with this.
>

I'm assuming Apple's snapshot program has an interface like:
```
$ make_TimeMachine_snapshot SOURCE
# ... Time Machine makes snapshot, and echoes

snapshot created in com.apple.TimeMachine.2019-02-23-102421
```

LVM, ZFS, and btrfs can do this.  Fedora, RHEL, and SUSE all default to
either LVM or btrfs.

I've lost mail due to IMAP IDLE PUSH notifying an email daemon (in this
case it was Akonadi) that new mail was available.  Akonadi started
downloading the mail, and as that was happening my scheduled backup
occurred.  After experiencing hardware failure the next day I tried to
restore my mail from backup and learned that some was missing (database
consistency issue).  IIRC this is how Apple Mail works too.

The two morals of this story for me were: 1) Don't trust email storage
in anything but a maildir, and/or 2) Always quiesce databases, then
snapshot the filesystem before making a backup; this insures
consistency.

I'm assuming that Time Machine is sane and quiesces databases controlled
by Apple software, and then makes an APFS snapshot before making the
Time Machine backup.

Vorta should not be inferior to Time Machine in insuring backup
consistency state.

>> I'm also not sure how often this ability would be used, but it seems like users who installed Vorta from PyPI might use it for things like activating the virtualenvironment.
>
> Not sure I understand what you mean. Wouldn't they have to activate
> the virtualenv before starting vorta in the first place? Or do you
> mean users who installed *borg* from PyPI?

Yes, thank you, that's what I meant :-)

> Anyway, I do agree power users will eventually come up with a use case for this and given your approach (script is always called with 2 parameters; default behaviour is to pass and quit), it really makes no sense to not call it on every operation.
>

Agreed.

> On the question of supported cases/subcommands: I suggest we just pass the subcommand to the script the same way it appears in vorta's (GUI) log (`check`/`list`/`prune`/`create`/`--version`). That way things should stay intact even if more subcommands become available. In the script I'd give examples for subcommands at the top and try to provide an extensive list, but always refer to borg and/or the vorta log for absolute certainty.
>

Agreed.  Also, I wonder if the example script should ideally contain a
sort of "hook API" check, because backwards and forwards compatibility
are not garanteed, and users could then diff the example script between
Vorta versions to see how they need to modify their script.

> On error handling: I suggest we handle errors the same way we do now. If any script call returns with a code ≠0, we abort there and cancel the operation. Anything else is too complicated and error-prone imho. Whether the script itself aborts on every error is then left to the user. They can do elaborate error-handling or just abort on the first error as well.
>

This makes sense for errors within the script.  For Borg or Vorta
errors, the script needs to be able to cleanup after itself.  For this
the hook API needs to support "cleanup" an argument to the script.

> I don't have a strong opinion on logging. Logging everything (stdout and stderr of the script and, consequently, of all programs called within) could clutter the log, but at the same time it would be the most user friendly option imho. So unless anyone disagree's, I'm going to prepare the PR to log everything. After all, users can always redirect to /dev/null in their script :)
>

:) Sounds good to me!

> I'll try to write a good example using POSIX SH and to make sure that users understand how it's going to be executed and what happens if they remove cases.
>

Thank you!

**Comment by sten0** [comment]
>> In addition to passing the subcommand as argument, it could pass more details as env vars. E.g. `VORTA_PROFILE_NAME`, `VORTA_REPO_URL`, ... Then the docs would have a sample script on how to deal with different subcommands?
>

I recommend avoiding env vars because of the potential issues they could
cause with virtualenvs, Flatpaks, and Snaps.  I expect that with
increasingly secure namespace barriers this will one day definitely
break.  This method also will need to be dropped when secure sudo/doas
support is (hopefully) one day added, because the normal user's env vars
must not be exported into the secure superuser (or more limited
CAP_ADMIN) environment.

> Isn't that already done for the current scripts?
> The hook name could also be passed as a env var so that the user can decide whether to pass it as an argument to the script.
>

What do you mean?

**Comment by real-yfprojects** [comment]
> What do you mean?

Below the pre- and post-backup command entries vorta states that the following env vars are available: `$repo_url, $profile_name, $profile_slug, $returncode`.

> I recommend avoiding env vars because of the potential issues they could cause with virtualenvs, Flatpaks, and Snaps. I expect that with increasingly secure namespace barriers this will one day definitely break. This method also will need to be dropped when secure sudo/doas support is (hopefully) one day added, because the normal user's env vars must not be exported into the secure superuser (or more limited CAP_ADMIN) environment.

Then we could allow the user to use placeholders in the entry where one defines the script/command.

**Comment by Erwyn** [comment]
> Well the thread is up to date, but unfortunately there's no PR yet. I have to admit I haven't looked into this for quite a while since I stopped doing backups on network mounts, but I will get back to this now. You can expect a PR in a few days and there we can have a final discussion on where to put the scripts and the other open questions.

Thanks for your answer, and sorry for the delay, was on vacation. I just wanted to be sure I wasn’t missing anything on my own setup.

thank you very much !

**Comment by real-yfprojects** [comment]
# Summary

* run backup only when hooks succeed
* [borgmatic hooks](https://torsion.org/borgmatic/docs/how-to/add-preparation-and-cleanup-steps-to-backups/)

## supported hooks

* run before a sequence of borg commands
* run after a sequence of borg commands
* before every borg command
* after every borg command
* before a specific command
* after a specific command
* on error
* 

## information for configured scripts
* archive name
* repo
* time?
* whether backup is scheduled
* hook
* sources
* borg arguments
* profile
* borg output?
* borg returncode

## supported actions

* mount/unmount a drive
* alter arguments / ENV variables for the following borg command
    * alter sources
    * alter archive name
* alter PATH / borg executable
* alter ssh command

## proposed GUI elements

* checkbox to toggle when to run configured scripts
* another entry for other hooks

### superior proposal
* run script for every hook 
* One entry, one script
* script decides what to do for the given hook
* script template or maybe generator?
    * in POSIX SH
    * security considerations with elevated rights
* option to disable script / all hooks\


## open questions

* How should the failure of an "after-borg" hook be handled?
* information for scripts passed as env variables or arguments?
* log script contents?
* restrict script location? - e.g. to Vorta's settings folder
* Allow specifying cmd arguments for the script?
* hook API check?
* what happens on borg error?
    * cleanup for pre script - also when pre script fails
* use env vars? - security concern in a root context
* How is information passed to scripts?
* Empose security restrictions on hooks like strict file permissions?
* options for how information is passed to script?
* logging inside the script: stderr, stdout
    * Error messages
    * warnings
    * info
    * debug

**Comment by AdwaitSalankar** [comment]
Here is what I thought the GUI should look like:
![Screenshot from 2024-03-24 00-54-06](https://github.com/borgbase/vorta/assets/111136306/7c60f512-ecdc-4c0f-aa0d-f6c080dc22ec)
Does it look good?

**Comment by real-yfprojects** [comment]
What's the difference between `Run script before and after all borg commands` and `Run script before and after every borg command`?

**Comment by m3nu** [comment]
Noticed the same. The UI is clearly not final and has logical issues.

I would look at this thread and see what use case most people want to solve. Then solve that and try to be flexible without making it too complex on the UI or the code. Maybe just 2 textboxes to run before `create` or `all` borg commands?

**Comment by AdwaitSalankar** [comment]
`before and after all borg commands` would mean before and after the complete workflow and `every borg command` would mean before and after each individual borg command in the workflow.
As mentioned in the summary, people want to mount and unmount the drive and alter arguments passed to Borg.

**Comment by Ultrabenosaurus** [comment]
I'm glad there's already a thread for this with a lot of good discussion. I would like to add my vote to this feature!

I am not using a network share, but a USB external drive that I don't want to auto-mount. Partly so I'm less tempted to use it for non-backup stuff, and partly so that in the unlikely event my machine gets scrambled or ransomwared, that drive and my backups should be untouched.

I also was surprised when the `prune`, `list`, and `check` actions failed after `create` finished due to the drive being unmounted already. I think the text in the GUI may need clarification, if this feature isn't going to be implemented soon. Specifically, on this _Shell Commands_ tab, the pre- and post-backup commands talk about the "backup" while the extra arguments field talks about "borg create".

![image](https://github.com/user-attachments/assets/9dee46bf-90d2-4329-84a3-8639f5140b54)

In my mind, this automatically translated into "the whole backup process" which would include `prune` and anything else that happens automatically with each backup run, scheduled or manual, versus "the creation of a new snapshot". If the text around the pre- and post-backup commands also stated it only applied to `borg create` then I would have been disappointed, but wouldn't have tried using them for mounting and unmounting my drive in the first place. I probably still would have ended up here requesting the feature, though :joy:

**Comment by ahmedhosssam** [comment]
I was trying to find a solution and I reached 2 approaches to solve it (based on people's feedback in the thread).

The first approach would be a GUI-based solution, where the user can go to the `advanced` section and add a script and choose its command from a checklist (in the same way in the video), and specify whether the script will run before or after these commands. The user can do this multiple times, they can import many scripts.

The second approach would be a minimal GUI with a single script and the script itself will specify all the information related to pre- and post-running, we can initialize a template script in a specific path that include the starting point, it will be something like this:

```
if [[ $1 == "pre_prune" ]]; then
    # execute pre-prune commands
fi

if [[ $1 == "post_prune" ]]; then
    # execute post-prune commands
fi

if [[ $1 == "pre_init" ]]; then
    # execute pre-init commands
fi
# ... and so on
```

And the user can modify it and add code for every if statement, we can let the user modify it from the path or from Vorta GUI.

The script itself will be run by the vorta code itself, we can modify `pre_post_backup_cmd` function to be `pre_post_cmd` and move it to a utils.py` file to be used by all borg commands, and each command will call the script with different arguments.

We can modify `pre_post_cmd` itself to run the template script and pass different arguments to it (e.g. pre_prune, post_prune, etc) based on the borg command.

We can pass more information to the script, but I'm not sure whether we pass it as env vars or arguments, we can do both? I don't know.

And of course all of this will be logged.

**Comment by ahmedhosssam** [comment]
Hi @m3nu ,
I'm planning to write my GSoC proposal on the `Advanced Borg Commands` idea, but I'm still not sure which way I should go with, the GUI with a bunch of buttons and menus to import external scripts, or just a single script containing to specify pre- and post information?

**Comment by m3nu** [comment]
I would do some mockups first and get user feedback

Personally, I think there should be a single line to enter very simple commands and a file selector to link to more complex scripts.

**Comment by frederikmoellers** [comment]
I prefer keeping the GUI as simple as possible and "hiding" advanced functionality in the scripts for power users. This lowers the entrance barriers for new users while still allowing advanced users to leverage the full functionality.

This would mean keeping the single script line as in the current UI and only possibly adding a checkbox "Run this script for every `borg` command" (as opposed to the default "Only run for `borg create`"). Setting the checkbox to "unchecked" by default also means we keep compatibility with existing configurations that expect the script to only run on `borg create`.

The script would either get a new parameter or a new environment variable containing the subcommand (and possibly another one for arguments?). Existing script that don't use that parameter/variable simply work as-is and new scripts could add functionality for every subcommand.

A template script as an example would show the possibilities using a `case` statement (for example).

Benefits of this solution:
- Keep GUI clean and unbloated
  - Don't increase the amount of information a new user has to work through before being able to use vorta
- Add support for most if not all possible use cases
  - The most prominent use cases described here (e.g. mounting a network share where backups are stored) are easy to implement
- Prevents users from configuring multiple scripts which might make debugging significantly harder if anything breaks

I honestly don't think getting feedback is going to help much. For one, the people who provide feedback will necessarily only represent a limited subset of the overall users which I think is unlikely to be representative. And it adds more work for the person trying to provide a solution. I would prefer getting feedback from a working solution and then having the option of improving that (possibly with the help of others, if they feel that the provided solution doesn't satisfy their needs).

But this is just my opinion. If you're the one solving the issue, it's your choice how you approach it 😉

**Comment by ahmedhosssam** [comment]
> I prefer keeping the GUI as simple as possible and "hiding" advanced functionality in the scripts for power users. This lowers the entrance barriers for new users while still allowing advanced users to leverage the full functionality.


This is my proposed idea for GSoC 2025 to solve the problem:

- Add a checkbox "Run script before and after Borg command" in the existing GUI.
- Create a new section called “Advanced Custom Hooks” as shown below.
- The new section is an advanced version of pre- and post-backup commands, it will be a text area containing a single script for pre- and post- scripts for all Borg commands.
- Also, this section might be hidden in a toggle button so it doesn’t confuse new users.
- When the user decides to use the advanced section, the default section will be deactivated, because it doesn't make sense to run both scripts for the same goal.

The end result for the UI will be something like this:

![Image](https://github.com/user-attachments/assets/3d108def-09de-4508-9594-4508e7c5486f)
